presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -292,14 +289,13 @@ public PlanWithProperties visitAggregation(AggregationNode node, Context context
                 return rebaseAndDeriveProperties(node, child);
             }
 
-            // aggregations would benefit from the finals being hash partitioned on groupId, however, we need to gather because the final HashAggregationOperator
-            // needs to know whether input was received at the query level.
-            if (node.getGroupingSets().stream().anyMatch(List::isEmpty)) {
+            if (node.hasEmptyGroupingSet() && !node.hasNonEmptyGroupingSet()) {
                 child = withDerivedProperties(
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
             else if (!child.getProperties().isStreamPartitionedOn(partitioningRequirement) && !child.getProperties().isNodePartitionedOn(partitioningRequirement)) {
+                // TODO: add check for decomposable functions?
```

Comment:
Is this a todo or is it not? If this is something that needs to be done, remove the ?